keps/sig-apps/20190412-pdb-for-scale-subresource.md

```diff
@@ -0,0 +1,84 @@
+---
+title: support-scale-subresource-in-pdbs
+authors:
+  - "@mortent"
+owning-sig: sig-apps
+participating-sigs:
+  - sig-scheduling
+  - sig-autoscaling
+reviewers:
+  - "@kow3ns"
+  - "@janetkuo"
+approvers:
+  - "@kow3ns"
+  - "@janetkuo"
+editor: TBD
+creation-date: 2019-04-12
+last-updated: 2019-04-12
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Support scale subresource in PDBs
+
+## Table of Contents
+
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+
+
+## Summary
+
+[Pod Disruption Budget (PDB)](https://kubernetes.io/docs/tasks/run-application/configure-pdb/)
+is a Kubernetes API that limits the number of pods of a collection that are down simultaneously from voluntary disruptions. PDBs allows a user to specify the allowed disruption through either min available or max unavailable number of pods. In order to support PDBs where max number of unavailable pods is set, the PDB controller needs to be able to look up the desired number of replicas. It does this by looking at the controller(s) (as specified by the owner ref) for the pods covered by the PDB. Currently only four of the basic workload controllers are supported for PDBs, namely Deployment, StatefulSet, ReplicaSet and ReplicaController. 
+
+The scale subresource allow any resource to specify its desired number of replicas and a generic way to look up this information. This document lays out a plan to use the scale subresource to allow setting PDBs for any resource implementing the scale subresource.
+
+## Motivation
+
+PDBs are an important tool to control the number of voluntary disruptions for workloads on Kubernetes. As more users start deploying custom controllers/operators based on CRDs, it is inconvenient that PDBs are not available. The scale subresource is already in use by the autoscaler and we should extend its use to PDBs.
+
+### Goals
+
+- Implement support for scale subresources in PDBs
+- Avoid major performance impact from the change
+
+## Proposal
+
+### Implementation Details/Notes/Constraints
+
+In the current implementation, the PDB controller identifies the workload controller by going through all the pods covered by the PDB, and for each pod, check in sequence for each of the four workload controllers. The PDB controller then looks up the desired number of replicas from the identified controller. If the set of pods covered by the PDB identfies more than one controller, that is considered an error. 
```

Comment:
Was it never updated after we added controller references? You can directly look up the controller by checking the object's ownerReference with the controller bit set. Querying N controller types is a super inefficient way to do this.

Reply:
It is using the controller references and only looks up a controller from the shared informers when the kind in the controllerref matches one of the kubernetes workload controllers. I have tried to make this clearer in the KEP.